optuna/storages/_in_memory.py

```diff
@@ -378,25 +380,31 @@ def _set_trial(self, trial_id: int, trial: FrozenTrial) -> None:
         study_id, trial_number = self._trial_id_to_study_id_and_number[trial_id]
         self._studies[study_id].trials[trial_number] = trial
 
-    def get_all_trials(self, study_id: int, deepcopy: bool = True) -> List[FrozenTrial]:
+    def get_all_trials(
+        self,
+        study_id: int,
+        deepcopy: bool = True,
+        state: Optional[Union[Tuple[TrialState, ...], TrialState]] = None,
+    ) -> List[FrozenTrial]:
 
         with self._lock:
             self._check_study_id(study_id)
-            if deepcopy:
-                return copy.deepcopy(self._studies[study_id].trials)
-            else:
-                return self._studies[study_id].trials[:]
 
-    def get_n_trials(self, study_id: int, state: Optional[TrialState] = None) -> int:
+            trials = self._studies[
+                study_id
+            ].trials  # type: Union[List[FrozenTrial], Iterator[FrozenTrial]]
```

Comment:
I'm not sure when we should use variable type annotation, but we can do here.

Reply:
Seemingly because `trials` is reassigned with a different type. I by the way don't think introducing new local variables will reduce the complexity here so I'd rather keep it as is.